File: java/src/main/java/year2015/day04/Year2015Day04.java
package year2015.day04;

import java.security.MessageDigest;

public class Year2015Day04 {

    public static long solve(String hashPrefix) throws Exception {
        String prefix = "iwrupvqb";
        long num = 0L;
        MessageDigest digest = MessageDigest.getInstance("MD5");
        while (true) {
            digest.reset();
            byte[] bytes = digest.digest((prefix + Long.toString(num)).getBytes());
            String hash = toHexString(bytes);

            if (hash.startsWith(hashPrefix)) {
                return num;
            }

            num++;
        }
    }

    // https://stackoverflow.com/questions/332079/in-java-how-do-i-convert-a-byte-array-to-a-string-of-hex-digits-while-keeping-l
    public static String toHexString(byte[] bytes) {
        StringBuilder hexString = new StringBuilder();

        for (byte b : bytes) {
            String hex = Integer.toHexString(0xFF & b);
            if (hex.length() == 1) {
                hexString.append('0');
            }
            hexString.append(hex);
        }

        return hexString.toString();
    }
}

File: js/src/year2015/day04/Year2015Day04.ts
const crypto = require('node:crypto');

export function solve(str: string, leading: number): number {
    let suffix = 0;
    outer: while (true) {
        const digest: string = crypto.createHash('md5')
            .update(str + suffix)
            .digest('hex');

        for (let i = 0; i < leading; i++) {
            if (digest.at(i) !== '0') {
                suffix++;
                continue outer;
            }
        }

        return suffix;
    }
}
